### ISSUE TITLE
FreeCAD installation

### ISSUE THREAD

**Issue by torodecamargue** [cadenasgmbh/3dfindit-freecad-integration#4]
Hello,
Really usefull application (see it works on video)
I try to install 3dfindit WB via the addon manager in FC 0.19 bit FC crashes each times I run the research on tour WB. 
I am on Windows 10. 

Can you help?

Julien

**Comment by berndhahnebach** [comment]
Which FreeCAD are you on?

Conda, LibPack, self compiled? Which exact version (commit) ?

When does it crash? When you switcht to the wb, when you click the search icon, when you try to start the search in the browser window? Do you see the browser task panel with the 3d find it search?

cheers bernd

**Comment by berndhahnebach** [comment]
BTW: works great on FreeCAD LibPack downloaded from here: https://github.com/FreeCAD/FreeCAD/releases/tag/0.19_pre

OS: Windows 10 (10.0)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.22960 (Git)
Build type: Release
Branch: master
Hash: c5a4b01d2e4218bcc0eb6650337650a6c65ef0e4
Python version: 3.6.8
Qt version: 5.12.1
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: German/Switzerland (de_CH)

**Comment by torodecamargue** [comment]
Hello,

Thank you for your answer. I use LibPack version (LP means LibPack I suppose)
FreeCAD_0.19.23074_Win-LPv12.1.6_vc14.x-x86-64.7z

It crashes when I click on the search icon.

Where can I copy paste data like you ?

**Comment by berndhahnebach** [comment]
The search icon on 3DfindIT websit in the FreeCAD TaskPanel or the FreeCAD icon which opens the 3DfindIT task panel? Would you try this in Python console?

```
from freecad.cadenas3dfindit import dialog
dialog.toggle()

```

- start FreeCAD
- Help --> About FreeCAD --> click on "Copy to clipboard"

- I just downloaded the latest available ... 

```
OS: Windows 10 (10.0)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.23074 (Git)
Build type: Release
Branch: master
Hash: 74b866ed3022fd1452c534dae31a721bcc668f2f
Python version: 3.6.8
Qt version: 5.12.1
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: German/Switzerland (de_CH)

```

**Comment by torodecamargue** [comment]
Hello,

I have try the Python code and it crashes

[url=https://zupimages.net/viewer.php?id=20/48/q9lr.jpg][img]https://zupimages.net/up/20/48/q9lr.jpg[/img][/url]

OS: Windows 10 (10.0)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.22988 (Git)
Build type: Release
Branch: master
Hash: cdc01cb20cfa74079c45b9dc2dab81e3248b2aa5
Python version: 3.6.8
Qt version: 5.12.1
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: French/France (fr_FR)

I see that I have an older version compared to you. Is it possible to upgrade the version without uninstall/ reinstall every WB that I already have ?

Julien

**Comment by torodecamargue** [comment]
Hello,

It crashes when I try the Python code.

I click on the serach icon on the left.

https://zupimages.net/viewer.php?id=20/48/q9lr.jpg

OS: Windows 10 (10.0)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.22988 (Git)
Build type: Release
Branch: master
Hash: cdc01cb20cfa74079c45b9dc2dab81e3248b2aa5
Python version: 3.6.8
Qt version: 5.12.1
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: French/France (fr_FR)


I have an older version of FC than you. Is it possible to upgrade the version without reinstall every WB already installed?

Julien

**Comment by berndhahnebach** [comment]
> It crashes when I try the Python code.
> 

on which command exactly? Line one or two?

If you installed them via AddOn manager or if you installed them all in your user Mod they all will stay there.

For a new FreeCAD just download the new FreeCAD, unzip and start FreeCAD exe, not even a need to remove the old one.

bernd

**Comment by n0x32schneider** [comment]
Simliar issue for me, when I run:

```
from freecad.cadenas3dfindit import dialog
dialog.toggle()

```

FreeCAD crashes. On the console I get the output:

```
[154216:154232:1207/012549.024329:ERROR:broker_posix.cc(40)] Recvmsg error: Connection reset by peer (104)
```

I built from current master

```
OS: Arch Linux
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.23237 (Git)
Build type: Release
Branch: master
Hash: 71726a11d0844f263dad636fb98bb996a479e007
Python version: 3.9.0
Qt version: 5.15.2
Coin version: 4.0.1
OCC version: 7.5.0
Locale: C/Default (C)
```

**Comment by berndhahnebach** [comment]
Would you help to find what exactly causes the problem?

When does it crash? on

`from freecad.cadenas3dfindit import dialog`

or on

`dialog.toggle()`

**Comment by n0x32schneider** [comment]
on `dialog.toggle()`, The dialog opens and does a loading animation (that loading bar) for a few seconds, then it crashes.

**Comment by berndhahnebach** [comment]
Great. First of all it makes sense to exact find wich of the commands in dialog.py does make FreeCAD crash. If we would know this, best to move with this information to the FreeCAD forum. 

BTW I neither have the problem on Windows nor on Linux.

**Comment by n0x32schneider** [comment]
When I visit the set Url directly in my browser (https://www.3dfindit.com/?portal=freecad-embedded&webview=QTWEBCHANNEL&lang=en_US) it shows the loading bar, but never does anything else.

Is that normal behaviour? If not it might be my browser.

**Comment by berndhahnebach** [comment]
same here on Debian Buster and firefox, never ending loading bar.

Would it be possible to test this code: 

```
from PySide2 import QtCore
from PySide2 import QtWidgets
import FreeCADGui
from freecad.cadenas3dfindit import dialog

browserWidget = QtWidgets.QDockWidget()
browserWidget.setObjectName("3Dfindit_Widget")
browserWidget.setWidget(dialog.Dialog())
FreeCADGui.getMainWindow().addDockWidget(QtCore.Qt.RightDockWidgetArea, browserWidget)

```

It works for me on Debian testing.

**Comment by n0x32schneider** [comment]
Code invokes same behaviour.
The widget opens, I see the loading bar, then it crashes FreeCAD.
Can I force freecad to write a logfile?

**Comment by berndhahnebach** [comment]
Both needs to be fixed, the crash and why it does not work. I would like to know if it the 3DfindIT website or the python code in the AddOn which causes the crash.

**Comment by berndhahnebach** [comment]
We may try to totally extract the code out of the AddOn to get it crashed without the AddOn at all.

**Comment by berndhahnebach** [comment]
@tsielaff Have you had any crash with the AddOn? I never ever had one.

**Comment by n0x32schneider** [comment]
What kind of safety mechanisms are there for python workbenches? How robust is the FreeCAD base instance if the workbench fails? 
Usually I get some kind of seg fault error, or something thrown in the console, when the base program crashes. But here there is nothing, everything just disappears.

**Comment by tsielaff** [comment]
No, I have not managed to make it crash. Neither on Windows nor on Linux.

**Comment by berndhahnebach** [comment]
I am not much involved in development of FreeCAD Base system. How about this: https://forum.freecadweb.org/viewtopic.php?f=10&t=25189&start=10#p422883 Does this code works for you?

```
import os
import tempfile
from urllib.request import urlretrieve
from PySide2 import QtCore
from PySide2 import QtWidgets
from PySide2 import QtWebEngineWidgets
from PySide2 import QtWebChannel

import FreeCAD
import FreeCADGui
import Part


the_html_code = """
<html>
<head>
    <script src="qrc:///qtwebchannel/qwebchannel.js"></script>
</head>
<header><title>title</title></header>
<body>

<p id="output"></p>

<h2><a href="#" onclick="jsWebPartImporter.text('https://forum.freecadweb.org/download/file.php?id=48016')">Click to import a Part from Internet into FreeCAD</a></h2>
<h2><a href="#" onclick="alert('Javascript works!')">Click for Java Script Test</a></h2>

<script type="text/javascript">
    window.onload = function() {
        new QWebChannel(qt.webChannelTransport, function (channel) {
            window.jsWebPartImporter = channel.objects.jsWebPartImporter;
            console.log(jsWebPartImporter);
        });
    }
</script> 
</body>
</html>
"""


class WebPartImporter(QtCore.QObject):
    def __init__(self, parent=None):
        super(WebPartImporter, self).__init__(parent)

    @QtCore.Slot(str)
    def text(self, url_to_step_file):

        print(url_to_step_file)

        # get a temp file with ending step
        tmpstepfile_path = tempfile.NamedTemporaryFile(suffix=".step").name
        print(tmpstepfile_path)

        # download the step file
        urlreturn = urlretrieve(url_to_step_file, tmpstepfile_path)
        print(urlreturn[0])

        # read the step (the read method needs file ending step)
        # and add it to the active document
        # create a new doc if there is no active doc
        Part.show(Part.read(tmpstepfile_path))
        os.remove(tmpstepfile_path)
        print(FreeCAD.ActiveDocument.Name)

        # set the view
        FreeCADGui.ActiveDocument.activeView().viewAxonometric()
        FreeCADGui.SendMsgToActiveView("ViewFit")


class BrowserWidget(QtWidgets.QDialog):
    def __init__(self):
        super(BrowserWidget, self).__init__()
        self.web_view = QtWebEngineWidgets.QWebEngineView()

        channel = QtWebChannel.QWebChannel(self)
        channel.registerObject("jsWebPartImporter", WebPartImporter(self.web_view))
        self.web_view.page().setWebChannel(channel)

        # layout 
        self.web_view.setSizePolicy(
            QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Expanding
        )
        self.layout = QtWidgets.QVBoxLayout(self)
        self.layout.addLayout(QtWidgets.QHBoxLayout())
        self.layout.addWidget(self.web_view)

    def showEvent(self, event):
        super(BrowserWidget, self).showEvent(event)
        self.web_view.setHtml(the_html_code)


mw = FreeCADGui.getMainWindow()
d = QtWidgets.QDockWidget()
d.setWidget(BrowserWidget())
mw.addDockWidget(QtCore.Qt.RightDockWidgetArea, d)


```

**Comment by berndhahnebach** [comment]
```
import queue
import tempfile
import threading
from concurrent.futures import ThreadPoolExecutor
from urllib.request import urlretrieve
from zipfile import ZipFile
from PySide2 import QtCore
from PySide2 import QtWidgets
from PySide2 import QtWebChannel
from PySide2 import QtWebEngineWidgets
import FreeCADGui


class JsTaskExecutor:
  class JsTask:
    def __init__(self, fn, order, args, kwargs):
      self.fn = fn
      self.order = order
      self.args = args
      self.kwargs = kwargs

    def __lt__(self, other):
      return self.order < other.order

    def __le__(self, other):
      return self.order <= other.order

    def __gt__(self, other):
      return self.order > other.order

    def __ge__(self, other):
      return self.order >= other.order

    def run(self):
      self.fn(*self.args, **self.kwargs)

  def __init__(self, threeDNativeAPI):
    # Save the native API object for later.
    self.threeDNativeAPI = threeDNativeAPI

    # We'll use a priority-based queue instead of the  normal queue.
    self.queue = queue.PriorityQueue()

    # Just an incrementing number to ensure all calls with the same
    # priority are done in the order they are placed.
    self.order = 0

    # Configure the thread that will do the actual work.
    self.thread = threading.Thread(target=self._executor, daemon=True)
    self.thread.start()

  def _executor(self):
    while True:
      # Wait. We need to make sure the API is ready before taking
      # a task. Any amount of time can pass while waiting for a task
      # so we need to make sure the API is still ready after we got a
      # task.
      self.threeDNativeAPI.isReady.wait()
      task = self.queue.get(block=True)[1]
      self.threeDNativeAPI.isReady.wait()

      # Run the task. There is still a slim chance that we run into
      # a reload very, very shortly before this task is run. Nothing
      # I can do...
      task.run()

  def submit(self, fn, *args, **kwargs):
      priority = kwargs.get('priority')
      if 'priority' in kwargs:
        del kwargs['priority']

      # Queue up the task and increment the order.
      self.queue.put((priority, JsTaskExecutor.JsTask(fn, self.order, args, kwargs)))
      self.order += 1


class MyAPI:
  def __init__(self, webView, three3DNativeAPI):
    self.webView = webView
    self.executor = JsTaskExecutor(three3DNativeAPI)

  def runJs(self, script, priority=50):
    self.executor.submit(self._runJs, script=script, priority=priority)

  def _runJs(self, script):
    self.webView.page().runJavaScript("(function(){ return " + script + ";})();")

  def loadByMident(self, mident):
    self.runJs("window.ThreeDfinditAPI.loadByMident('" + mident + "')")

  def loadByMidentBase64(self, midentBase64):
    self.runJs("window.ThreeDfinditAPI.loadByMidentBase64('" + midentBase64 + "')")

  def loadByIDStr(self, idStr):
    self.runJs("window.ThreeDfinditAPI.loadByIDStr('" + idStr + "')")

  def loadByIDStrBase64(self, idStrBase64):
    self.runJs("window.ThreeDfinditAPI.loadByIDStrBase64('" + idStrBase64 + "')")

  def setProperty(self, prop, value):
    self.runJs("window.ThreeDfinditAPI.setProperty('" + prop + "', '" + value + "')", 1)

  def startGeoSearch(self, filename):
    self.runJs("window.ThreeDfinditAPI.startGeoSearch('" + filename + "')")

  def sendGeoSearchChunkBase64(self, chunk):
    self.runJs("window.ThreeDfinditAPI.sendGeoSearchChunkBase64('" + chunk + "')")

  def endGeoSearchChunkBase64(self):
    self.runJs("window.ThreeDfinditAPI.endGeoSearchChunkBase64()")

  def doGeoSearch(self, openSearchResults):
    self.runJs("window.ThreeDfinditAPI.doGeoSearch(" + str(openSearchResults).lower() + ")")

  def doSketchSearch(self, front, top, side, type, openSearchResults):
    self.runJs("window.ThreeDfinditAPI.doSketchSearch('" + front + "', '" + top + "', '" + side + "', '" + type + "', " + str(openSearchResults).lower() + ")")


class MyNativeAPI(QtCore.QObject):
  def __init__(self, webView):
    super(MyNativeAPI, self).__init__(webView)

    # Add an event that is fired when the API becomes ready to use.
    self.isReady = threading.Event()

  @QtCore.Slot()
  def ready(self):
    self.isReady.set()

  @QtCore.Slot("QJsonObject")
  def downloadReadyObject(self, downloadReadyObj):
    if downloadReadyObj["isExternal"]:
      # Alert user that this is a crawled document.
      msgBox = QtWidgets.QMessageBox()
      msgBox.setIcon(QtWidgets.QMessageBox.Question)
      msgBox.setWindowTitle("3DfindIT.com")
      msgBox.setText("Should 3DfindIT redirect you to the website of the supplier?")
      msgBox.setStandardButtons(QtWidgets.QMessageBox.Yes | QtWidgets.QMessageBox.No)
      msgBox.setWindowModality(QtCore.Qt.ApplicationModal)
      if (msgBox.exec_() == QtWidgets.QMessageBox.Yes):
        # Open in system browser.
        webbrowser.open(downloadReadyObj["url"], new=2)        
    else:
      # Download file.
      tmpDownloadPath = tempfile.mkstemp(prefix="3df", suffix=".zip")[1]
      urlretrieve(downloadReadyObj["url"], tmpDownloadPath)

      # Extract zip file.
      tmpExtractPath = tempfile.mkdtemp(prefix="3df")
      with ZipFile(tmpDownloadPath, 'r') as zip:
          zip.extractall(tmpExtractPath)

      # Open the STEP file.
      stepFile = tmpExtractPath + "\\" + downloadReadyObj["startFile"]
      Part.show(Part.read(stepFile), downloadReadyObj["NB"])

      # Fit to view.
      FreeCADGui.SendMsgToActiveView("ViewFit")


class MyWebEnginePage(QtWebEngineWidgets.QWebEnginePage):
  def certificateError(self, error):
    # Ask the user if we should proceed with an invalid certifacate.
    msgBox = QtWidgets.QMessageBox()
    msgBox.setText("An invalid certificate was encountered while loading 3DFindIT. Proceed anyway?")
    msgBox.setStandardButtons(QtWidgets.QMessageBox.Yes | QtWidgets.QMessageBox.No)
    msgBox.setWindowModality(QtCore.Qt.ApplicationModal)
    if (msgBox.exec_() == QtWidgets.QMessageBox.Yes):
      return True
    return False


class MyDialog(QtWidgets.QDialog):
  def __init__(self):
    super(MyDialog, self).__init__()

    self.webView = QtWebEngineWidgets.QWebEngineView()
    self.webView.setSizePolicy(QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Expanding)

    # Init native API.
    self.threeDNativeAPI = MyNativeAPI(self.webView)

    # Init API.
    self.threeDAPI = MyAPI(self.webView, self.threeDNativeAPI)
    self.threeDAPI.setProperty("cadsystem", "freecad")
    self.threeDAPI.setProperty("cadversion", "0.19")
    self.threeDAPI.setProperty("productname", "FreeCAD")

    self.layout = QtWidgets.QVBoxLayout(self)
    self.layout.addLayout(QtWidgets.QHBoxLayout())
    self.layout.addWidget(self.webView)

    # Setup our page. We want our own profile so other browsers in FreeCAD
    # can't interfere with our cookies. We also want our own page so we can
    # ask the user what to do if we encounter any SSL errors.
    profile = QtWebEngineWidgets.QWebEngineProfile("3Dfindit", self.webView)
    page = MyWebEnginePage(QtWebEngineWidgets.QWebEnginePage(profile, self.webView))
    self.webView.setPage(page)

    channel = QtWebChannel.QWebChannel(self)
    channel.registerObject("ThreeDfinditNativeAPI", self.threeDNativeAPI)
    self.webView.page().setWebChannel(channel)
    self.webView.setUrl("https://freecad-embedded.3dfindit.com/?webview=QTWEBCHANNEL")


browserWidget = QtWidgets.QDockWidget()
browserWidget.setWidget(MyDialog())
FreeCADGui.getMainWindow().addDockWidget(QtCore.Qt.RightDockWidgetArea, browserWidget)

```

**Comment by n0x32schneider** [comment]
[https://i.imgur.com/AJX8ltw.png](url)
First snippet works and does not crash.

Second one does crash like before.

**Comment by berndhahnebach** [comment]
https://forum.freecadweb.org/viewtopic.php?f=10&t=53157

**Comment by n0x32schneider** [comment]
I rebuilt from current master, and now I get some new error on the console:
`
[64921:64949:1211/181644.616507:ERROR:service_worker_storage.cc(1575)] Failed to delete the database: Database IO error
`

The loading animation lasts longer now. The program does not terminate as soon as the error is thrown, but a few seconds later.

I also tried replacing the url with another url. 
Opening google in the browserwidget works without crash.
